lib/services/ngsi/subscriptionService.js

```diff
@@ -129,22 +182,99 @@ function subscribe(device, triggers, content, callback) {
     if (device.cbHost) {
         options.uri = 'http://' + device.cbHost + '/v1/subscribeContext';
     } else {
-        options.uri = 'http://' + config.getConfig().contextBroker.host + ':' +
-            config.getConfig().contextBroker.port + '/v1/subscribeContext';
+        options.uri = config.getConfig().contextBroker.url + '/v1/subscribeContext';
     }
 
-    request(options, createSubscriptionHandler(device, triggers, store, callback));
+    request(options, createSubscriptionHandlerNgsi1(device, triggers, store, callback));
 }
 
 /**
- * Generate a new unsubscription request handler, based on the device and subscription ID given to the function.
+ * Makes a subscription for the given device's entity using NGSIv2, triggered by the given attributes.
+ * The contents of the notification can be selected using the "content" array (that can be left blank
+ * to notify the complete entity).
+ *
+ * @param {Object} device       Object containing all the information about a particular device.
+ * @param {Object} triggers     Array with the names of the attributes that would trigger the subscription
+ * @param {Object} content      Array with the names of the attributes to retrieve in the notification.
+ */
+function subscribeNgsi2(device, triggers, content, callback) {
+    var options = {
+        method: 'POST',
+        headers: {
+            'fiware-service': device.service,
+            'fiware-servicepath': device.subservice
+        },
+        json: {
+                subject:
+                {
+                    entities: [
+                        {
+                            id: device.name,
+                            type: device.type
+                        }
+                    ],
+
+                    condition: {
+                        attrs: triggers
+                    },
+                },
+                notification: {
+                    http: {
+                        url: config.getConfig().providerUrl + '/notify'
+                    },
+                    attrs: content || [],
+                    attrsFormat: 'normalized'
+                },
+                throttling: moment.duration(config.getConfig().throttling).seconds()
+        }
+    };
+
+    if (config.getConfig().deviceRegistrationDuration) {
```

Comment:
After running e2e integration test I think there is an issue in the way of calculating expiration in NGSIv2 subscription logic.

Note that `config.getConfig().deviceRegistrationDuration` is a string in ISO8601 duration format, such as `P1M` (see https://github.com/telefonicaid/iotagent-node-lib/blob/master/config.js#L73).

I'm not an expert, but as far as I understand this format is not compatible with the ones used by  [add() function](https://momentjs.com/docs/#/manipulating/add/) and I have check empirically that `.add(config.getConfig().deviceRegistrationDuration)` adds 0, so the subscription expires instantaneously and makes test cases related with reverse bidirectionality plugin to fail. Moreover, if I hack this part of the code, e.g. `options.json.expires = 1841595324`  (a date in 2028) then the tests work again.

Is this makes sense to you (I may be wrong :) this should be fixed (maybe with a function to adapt the ISO8601 duration format to the one that adds expects). In addition, we should have a look to unit test to discover why this has been passed.

Reply:
I have reviewed the functionality and you are completely right. There is a small bug in the subscription functionality and in the corresponding unit tests.

Reply:
Unit tests fixed with 9515f31fa8d1cae5fa4abffc6ad7cba993f55fea. As you can see now they are detecting the error.